controlplane/kubeadm/internal/controllers/remediation.go

```diff
@@ -168,6 +169,23 @@ func (r *KubeadmControlPlaneReconciler) reconcileUnhealthyMachines(ctx context.C
 	// Before starting remediation, run preflight checks in order to verify it is safe to remediate.
 	// If any of the following checks fails, we'll surface the reason in the MachineOwnerRemediated condition.
 
+	if feature.Gates.Enabled(feature.ClusterTopology) {
+		// Skip remediation when we expect an upgrade to be propagated from the cluster topology.
+		if controlPlane.Cluster.Spec.Topology != nil && controlPlane.Cluster.Spec.Topology.Version != controlPlane.KCP.Spec.Version {
+			log.Info("A control plane machine needs remediation, but waiting for a pending version upgrade. Skipping remediation", "version", controlPlane.Cluster.Spec.Topology.Version)
+			conditions.MarkFalse(machineToBeRemediated, clusterv1.MachineOwnerRemediatedCondition, clusterv1.WaitingForRemediationReason, clusterv1.ConditionSeverityWarning, "KubeadmControlPlane can't remediate if a version upgrade is pending")
+
+			v1beta2conditions.Set(machineToBeRemediated, metav1.Condition{
+				Type:    clusterv1.MachineOwnerRemediatedV1Beta2Condition,
+				Status:  metav1.ConditionFalse,
+				Reason:  controlplanev1.KubeadmControlPlaneMachineCannotBeRemediatedV1Beta2Reason,
+				Message: "KubeadmControlPlane can't remediate if a version upgrade is pending",
+			})
+
+			return ctrl.Result{}, nil
```

Comment:
Wondering if this is correct here.

We will get another reconcile if KCP.spec.version changes, but we won't get another one if Cluster.spec.topology.version changes (e.g. if someone sets back that version to KCP.spec.version).

But maybe instead of changing the Result here we should change the filter predicate for Cluster events?

Reply:
> But maybe instead of changing the Result here we should change the filter predicate for Cluster events?

Do you mean having Or (ClusterPausedTransitionsOrInfrastructureReady, VersionIsChanged) or something similar

Reply:
Yes, exactly. We'll add that